src/expr/src/scalar/func.rs

```diff
@@ -3626,10 +3752,25 @@ impl UnaryFunc {
             CharLength => char_length(a),
             IsRegexpMatch(regex) => Ok(is_regexp_match_static(a, &regex)),
             RegexpMatch(regex) => regexp_match_static(a, temp_storage, &regex),
-            DatePartInterval(units) => date_part_interval_inner(*units, a.unwrap_interval()),
-            DatePartTime(units) => date_part_time_inner(*units, a.unwrap_time()),
-            DatePartTimestamp(units) => date_part_timestamp_inner(*units, a.unwrap_timestamp()),
-            DatePartTimestampTz(units) => date_part_timestamp_inner(*units, a.unwrap_timestamptz()),
+            ExtractInterval(units) => {
+                date_part_interval_inner::<Numeric>(*units, a.unwrap_interval())
```

Comment:
This was not your doing, but all of these unwraps could just get moved inside this inner func:

```diff
diff --git a/src/expr/src/scalar/func.rs b/src/expr/src/scalar/func.rs
index 331e1acdb..683d941d6 100644
--- a/src/expr/src/scalar/func.rs
+++ b/src/expr/src/scalar/func.rs
@@ -1735,13 +1735,11 @@ fn date_part_interval(a: Datum, interval: Interval) -> Result<Datum, EvalError>
     }
 }
 
-fn date_part_interval_inner<T>(
-    units: DateTimeUnits,
-    interval: Interval,
-) -> Result<Datum<'static>, EvalError>
+fn date_part_interval_inner<T>(units: DateTimeUnits, a: Datum) -> Result<Datum<'static>, EvalError>
 where
     T: DecimalLike + Into<Datum<'static>>,
 {
+    let interval = a.unwrap_interval();
     match units {
         DateTimeUnits::Epoch => Ok(interval.as_seconds::<T>().into()),
         DateTimeUnits::Millennium => Ok(T::from(interval.millennia()).into()),
@@ -3752,9 +3750,7 @@ impl UnaryFunc {
             CharLength => char_length(a),
             IsRegexpMatch(regex) => Ok(is_regexp_match_static(a, &regex)),
             RegexpMatch(regex) => regexp_match_static(a, temp_storage, &regex),
-            ExtractInterval(units) => {
-                date_part_interval_inner::<Numeric>(*units, a.unwrap_interval())
-            }
+            ExtractInterval(units) => date_part_interval_inner::<Numeric>(*units, a),
             ExtractTime(units) => date_part_time_inner::<_, Numeric>(*units, a.unwrap_time()),
             ExtractTimestamp(units) => {
                 date_part_timestamp_inner::<_, Numeric>(*units, a.unwrap_timestamp())
@@ -3763,7 +3759,7 @@ impl UnaryFunc {
                 date_part_timestamp_inner::<_, Numeric>(*units, a.unwrap_timestamptz())
             }
             ExtractDate(units) => extract_date_inner(*units, a.unwrap_date()),
-            DatePartInterval(units) => date_part_interval_inner::<f64>(*units, a.unwrap_interval()),
+            DatePartInterval(units) => date_part_interval_inner::<f64>(*units, a),
             DatePartTime(units) => date_part_time_inner::<_, f64>(*units, a.unwrap_time()),
             DatePartTimestamp(units) => {
                 date_part_timestamp_inner::<_, f64>(*units, a.unwrap_timestamp())
```

note that I don't think this blithely works with some of the other suggestions

Reply:
Just pushed this change for `interval`, `time`, and `date`.